components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlinx/coroutines/flow/Flow.kt

```diff
@@ -47,6 +50,62 @@ fun <T, R> Flow<T>.ifChanged(transform: (T) -> R): Flow<T> {
     }
 }
 
+/**
+ * Returns a [Flow] containing only changed elements of the lists of the original [Flow].
+ *
+ * ```
+ * Example: Identity function
+ * Transform: x -> x (transformed values are the same as original)
+ * Original Flow: list(0), list(0, 1), list(0, 1, 2, 3), list(4), list(5, 6, 7, 8)
+ * Transformed:
+ * (0)          -> (0 emitted because no previous value at position)
+ *
+ * (0, 1)       -> (0 not emitted because same as previous value at position,
+ *                  1 emitted because no previous value at position),
+ *
+ * (0, 1, 2, 3) -> (0 and 1 not emitted because same as previous values,
+ *                  2 and 3 emitted because no previous values at positions),
+ *
+ * (4)          -> (4 emitted because different than previous value at position)
+ *
+ * (5, 6, 7, 8) -> (5, 6, 7, 8 all emitted because different than previous values at positions)
+ * Returned Flow: 0, 1, 2, 3, 4, 5, 6, 7, 8
+ * ---
+ *
+ * Example: Modulo 2
+ * Transform: x -> x % 2 (emit changed values if the result of modulo 2 changed)
+ * Transformed:
+ * (1)          -> (1 emitted because no previous mapped value at position)
+ *
+ * (1, 0)       -> (1 not emitted because same as previous mapped value at position,
+ *                  2 emitted because no previous mapped value at position),
+ *
+ * (1, 0, 1, 0) -> (5 and 6 not emitted because mapped values same as previous values,
+ *                  3 and 4 emitted because no previous mapped values at positions),
+ *
+ * (1)          -> (7 not emitted because mapped value same as previous mapped value at position)
+ *
+ * Returned Flow: 1, 2, 3, 4
+ * ---
+ * ```
+ */
+@FlowPreview
+fun <T, R> Flow<List<T>>.filterChanged(transform: (T) -> R): Flow<T> {
+    var lastMappedValues: List<R>? = null
+    return flatMapConcat { values ->
+        val lastMapped = lastMappedValues
+        val changed = if (lastMapped == null) {
+            values
+        } else {
+            values.filterIndexed { index, value ->
+                lastMapped.size <= index || lastMapped[index] !== transform(value)
+            }
```

Comment:
@pocmo yes, I think we need to care about this, because otherwise a tab could get added and eventually have a window request `w3` (or any other change we're observing) and we wouldn't notify it because the value wouldn't have changed relative to our last mapped values. So, if there's no value at the index of previous last mapped values, we have to emit as well.

Reply:
OK, pushed another update now!